psi4/src/psi4/libmints/twobody.cc

```diff
@@ -112,6 +116,61 @@ TwoBodyAOInt::TwoBodyAOInt(const TwoBodyAOInt &rhs) : TwoBodyAOInt(rhs.integral_
 
 TwoBodyAOInt::~TwoBodyAOInt() {}
 
+// Haser 1989, Equation 7 
+void TwoBodyAOInt::update_density(const std::vector<SharedMatrix>& D) {
+    
+    timer_on("Update Density");
+#pragma omp parallel for
+    for (int M = 0; M < nshell_; M++) {
+        for (int N = 0; N < nshell_; N++) {
+            int m_start = bs1_->shell(M).function_index();
+            int num_m = bs1_->shell(M).nfunction();
+
+            int n_start = bs1_->shell(N).function_index();
+            int num_n = bs1_->shell(N).nfunction();
+            
+            double max_dens = 0.0;
+            for (int m = m_start; m < m_start + num_m; m++) {
+                for (int n = n_start; n < n_start + num_n; n++) {
+                    double val = 0.0;
+                    for (int i = 0; i < D.size(); i++) {
+                        val += std::abs(D[i]->get(m, n));
+                    }
+                    if (val > max_dens) max_dens = val;
+                }
+            }
+            
+            max_dens_shell_pair_[M * nshell_ + N] = max_dens;
+        }
+    }
+    timer_off("Update Density");
+
+}
+
+// Haser 1989 Equations 6 to 14
+bool TwoBodyAOInt::shell_significant_density(int M, int N, int R, int S) {
+
+    // THR in Equation 9
+    double density_threshold = screening_threshold_squared_;
+
+    // Equation 13
+    double Q_MN_sq = shell_pair_values_[N * nshell_ + M];
+    double Q_RS_sq = shell_pair_values_[S * nshell_ + R];
+
+    std::initializer_list<double> block_densities {max_dens_shell_pair_[M * nshell_ + N], max_dens_shell_pair_[R * nshell_ + S], 
+        0.25 * max_dens_shell_pair_[M * nshell_ + R], 0.25 * max_dens_shell_pair_[M * nshell_ + S],
+        0.25 * max_dens_shell_pair_[N * nshell_ + R], 0.25 * max_dens_shell_pair_[N * nshell_ + S]};
+
+    // Equation 6
+    double max_density = std::max(block_densities);
+    // Squared to account for the fact that Q_MN is given as its square
+    max_density = max_density * max_density;
```

Comment:
Reading the paper, this is a major problem in your implementation. Haser's equation is that:
ε > Q_MN * Q_RS * ΔD.

You do not have his equation 13 correct. What you call Q is what Haser calls Q squared. To "correct" for this, you square ΔD, but then you would need to square ε as well. So when the user inputs a screening threshold, they are effectively getting the square root of that threshold. That's bad.

The simpler thing to do is just define Q as Haser does.

Reply:
Ah, I set the density threshold to the square of the screening threshold, so my code was still valid, let me change it to make it more clear :) Schwarz Screening also does this, to avoid taking expensive square roots.

Reply:
I'm going to request a second opinion from somebody more knowledgeable about screening procedures (@andysim?). What bothers me is that `screening_threshold_` is determined by `INTS_TOLERANCE`, and the documentation on `INTS_TOLERANCE` mentions nothing about this square/square-root behavior. At the very least, this is undocumented behavior.